File: android/app/src/main/java/com/cnezsoft/zentao/ZentaoAPI.java
package com.cnezsoft.zentao;

import com.cnezsoft.zentao.data.EntryType;

import org.json.JSONException;
import org.json.JSONObject;

import java.math.BigInteger;
import java.net.MalformedURLException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

/**
 * Zentao API helpers
 * Created by Catouse on 2015/1/12.
 */
public class ZentaoAPI
{
    public enum RequestType {
        GET,
        PATH_INFO
    }

    public static final boolean GZIP_REQUEST = true;

    /**
     * Encrypted string with 'md5' algorithm
     * @param s
     * @return
     */
    public static String md5(String s)
    {
        try
        {
            MessageDigest digest = MessageDigest.getInstance("MD5");
            digest.update(s.getBytes(), 0, s.length());
            String hash = new BigInteger(1, digest.digest()).toString(16);
            return hash;
        }
        catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        return "";
    }

    /**
     * Concat zentao api url with params
     * @param params
     * @param user
     * @return
     */
    public static String concatUrl(Map<String, String> params, User user)
    {
        ZentaoConfig config = user.getZentaoConfig();
        String moduleName = params.get("module");
        String methodName = params.get("method");
        String viewType = Helper.ifNullOrEmptyThen(params.get("viewType"), "json");
        String password;
        String url = user.getAddress();
        if(!url.endsWith("/")) url += "/";

        if(config.getRequestType() == RequestType.PATH_INFO)
        {
            if(moduleName.equals("user") && methodName.equals("login"))
            {
                password = md5(user.getPasswordMd5() + config.getRand());
                url += "user-login." + viewType
                    + "?account=" + user.getAccount()
                    + "&password=" + password
                    + "&" + config.getSessionName() + "=" + config.getSessionID();
                return url;
            }

            url += moduleName + "-" + methodName + "-";

            if(moduleName.equals("api"))
            {
                if(methodName.toLowerCase().equals("mobilegetlist"))
                {
                    url += Helper.ifNullOrEmptyThen(params.get("type"), "full") + "-"
                        + Helper.ifNullOrEmptyThen(params.get("object"), "all") + "-"
                        + Helper.ifNullOrEmptyThen(params.get("range"), "0") + "-"
                        + Helper.ifNullThen(params.get("last"), "") + "-"
                        + Helper.ifNullOrEmptyThen(params.get("records"), "1000") + "-"
                        + Helper.ifNullOrEmptyThen(params.get("format"), "index");
                }
                else if(methodName.toLowerCase().equals("mobilegetinfo"))
                {
                    url += params.get("id") + "-" + params.get("type");
                }
            }

            if(url.endsWith("-"))
            {
                url = url.substring(0, url.length() - 1);
            }

            url += "." + viewType + "?" + config.getSessionName() + "=" + config.getSessionID();
        }
        else // config.getRequestType() == RequestType.GET
        {
            url += "/index.php?";
            if(moduleName.equals("user") && methodName.equals("login"))
            {
                password = md5(user.getPasswordMd5() + config.getRand());
                url += "m=user&f=login&account=" + user.getAccount() +
                        "&password=" + password + "&" + config.getSessionName() +
                        "=" + config.getSessionID() + "&t=" + viewType;
                return url;
            }

            url += "m=" + moduleName + "&f=" + methodName;

            if(moduleName.equals("api"))
            {
                if(methodName.toLowerCase().equals("mobilegetlist"))
                {
                    url += "&type=" + Helper.ifNullOrEmptyThen(params.get("type"), "full")
                        + "&object=" + Helper.ifNullOrEmptyThen(params.get("object"), "all")
                        + "&range=" + Helper.ifNullOrEmptyThen(params.get("range"), "0")
                        + "&last=" + Helper.ifNullThen(params.get("last"), "")
                        + "&records=" + Helper.ifNullOrEmptyThen(params.get("records"), "all")
                        + "&format=" + Helper.ifNullOrEmptyThen(params.get("format"), "index");
                }
                else if(methodName.toLowerCase().equals("mobilegetinfo"))
                {
                    url += "&id=" + params.get("id")
                        + "&type=" + params.get("type");
                }
            }

            url += "&" + config.getSessionName() + "=" + config.getSessionID();
        }
        return url;
    }

    /**
     * Get zentao config
     * The method should called in async task
     *
     * @param userAddress
     * @return
     * @throws MalformedURLException
     * @throws JSONException
     */
    public static ZentaoConfig getConfig(String userAddress) {
        JSONObject jsonConfig = Http.getJSON(userAddress + "/index.php?mode=getconfig");
        if(jsonConfig == null) return null;
        return new ZentaoConfig(jsonConfig);
    }

    /**
     * Login in zentao
     * The method should called in async task
     *
     * @param user
     * @return
     */
    public static OperateBundle<Boolean, User> login(User user)
    {
        if(!user.hasLoginCredentials()) {
            return new OperateBundle<Boolean, User>(false, "User information required.", user) {{setCode(5);}};
        }

        boolean result;
        String message;
        Map<String, String> params = new HashMap<String, String>(){{put("module", "user"); put("method", "login");}};
        JSONObject jsonResult = Http.getJSON(concatUrl(params, user));

        if(jsonResult != null) {
            String status = jsonResult.optString("status", "failed");
            result = status.equals("success");
            message = jsonResult.optString("reason");

            if(result) {
                try {
                    JSONObject jsonUser = jsonResult.getJSONObject("user");
                    user.fromJSON(jsonUser);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            final int code = result ? 0 : 2;
            return new OperateBundle<Boolean, User>(result, message, user) {{setCode(code);}};
        } else {
            return new OperateBundle<Boolean, User>(false, "Can't get data from remote server.", user) {{setCode(4);}};
        }
    }

    /**
     * Get zentao config then login in zentao
     * The method should called in async task
     *
     * @param user
     * @return
     */
    public static OperateBundle<Boolean, User> tryLogin(User user) {
        if(!user.hasLoginCredentials()) {
            return new OperateBundle<Boolean, User>(false){{setCode(5);}};
        }

        ZentaoConfig config = getConfig(user.getAddress());
        user.setZentaoConfig(config);

        // Check zentao version
        if(config == null) {
            // Can't connect to the zentao server.
            return new OperateBundle<Boolean, User>(false){{setCode(1);}};
        }
        else if(!config.isPro() || config.getVersionNumber() < 4.3f ) {
            // Zentao version is not correct
            return new OperateBundle<Boolean, User>(false, user){{setCode(3);}};
        }

        return login(user);
    }

    public static OperateBundle<Boolean, JSONObject> getDataItem(User user, EntryType entryType, String id) {
        Map<String, String> parmas = new HashMap<>();
        parmas.put("module", "api");
        parmas.put("method", "mobileGetInfo");
        parmas.put("type", entryType.name().toLowerCase());
        parmas.put("id", id);

        JSONObject json = Http.getJSON(concatUrl(parmas, user), GZIP_REQUEST);
        boolean result;
        String message;
        if(json != null) {
            String status = json.optString("status", "failed");
            result = status.equals("success");
            message = json.optString("reason");

            if(result) {
                json = json.optJSONObject("data");
                if(json == null && Helper.isNullOrEmpty(message)) {
                    message = "No new data. ";
                }
            }
        } else {
            result = false;
            message = "Cant' get data from remote server.";
        }
        return new OperateBundle<>(result, message, json);
    }

    /**
     * Get data list
     * @param user
     * @param type
     * @param entryType
     * @param range
     * @param records
     * @param format
     * @return
     */
    public static OperateBundle<Boolean, JSONObject> getDataList(User user, String type, EntryType entryType,
        int range, int records, String format) {
        JSONObject json;
        boolean result;
        String message;
        Map<String, String> parmas = new HashMap<>();
        Date lastSyncTime = user.getLastSyncTime();
        parmas.put("module", "api");
        parmas.put("method", "mobileGetList");
        parmas.put("type", type);
        parmas.put("object", entryType == EntryType.Default ? "all" : entryType.name().toLowerCase());
        parmas.put("range", range + "");
        parmas.put("last", ((int) Math.floor(lastSyncTime.getTime()/1000)) + "");
        parmas.put("records", records + "");
        parmas.put("format", format);

        json = Http.getJSON(concatUrl(parmas, user), GZIP_REQUEST);
        if(json != null) {
            String status = json.optString("status", "failed");
            result = status.equals("success");
            message = json.optString("reason");

            if(result) {
                json = json.optJSONObject("data");
                if(json == null && Helper.isNullOrEmpty(message)) {
                    message = "No new data. last sync time=" + lastSyncTime.toString();
                } else {
                    message = "last sync time=" + lastSyncTime.toString();
                }
            }
        }
        else {
            result = false;
            message = "Cant' get data from remote server.";
        }
        return new OperateBundle<>(result, message, json);
    }

    /**
     * Get data list
     * @param user
     * @param entryType
     * @return
     */
    public static OperateBundle<Boolean, JSONObject> getDataList(User user, EntryType entryType) {
        return getDataList(user, "increment", entryType, 0, 1000, "index");
    }

    /**
     * Get data list
     * @param user
     * @return
     */
    public static OperateBundle<Boolean, JSONObject> getDataList(User user) {
        return getDataList(user, "increment", EntryType.Default, 0, 1000, "index");
    }
}

File: iOS/zentao/zentao/ZentaoAPI.swift
//
//  ZentaoAPI.swift
//  zentao
//
//  Created by Sun Hao on 15/3/12.
//  Copyright (c) 2015年 cnezsoft.com. All rights reserved.
//

import Foundation

struct ZentaoAPI {

    /// Concat api url
    static func concatUrl(params: [String:String], withUser user: User) -> String {
        let moduleName = params["module"] ?? "api"
        let methodName: String = params["method"]!
        let moduleNameIdentifier = moduleName.lowercaseString
        let methodNameIdentifier = methodName.lowercaseString
        let viewType = params["viewType"] ?? "json"
        var url = user.address.hasSuffix("/") ? user.address : user.address + "/";
        
        if(user.config!.isPathInfoRequestType) {
            if moduleNameIdentifier == "user" && methodNameIdentifier == "login" {
                url += "user-login.\(viewType)"
                    + "?account=\(user.account)&password=\(user.passwordMD5WithRand)"
                    + "&\(user.config!.sessionName!)=\(user.config!.sessionID!)"
                return url;
            }
            
            url += "\(moduleName)-\(methodName)-"
            
            if moduleNameIdentifier == "api" {
                if methodNameIdentifier == "mobilegetlist" {
                    url += (params["type"] ?? "full") + "-"
                    url += (params["object"] ?? "all") + "-"
                    url += (params["range"] ?? "0") + "-"
                    url += (params["last"] ?? "") + "-"
                    url += (params["records"] ?? "1000") + "-"
                    url += (params["format"] ?? "index")
                } else if methodNameIdentifier == "mobilegetinfo" {
                    url += params["id"]! + "-" + params["type"]!
                }
            }
            
            if url.hasSuffix("-") {
                url = (url as NSString).substringToIndex(countElements(url) - 1)
            }
            
            url += ".\(viewType)?\(user.config!.sessionName!)=\(user.config!.sessionID!)"
        } else {
            url += "/index.php?"
            if moduleNameIdentifier == "user" && methodNameIdentifier == "login" {
                url += "m=user&f=login&account=\(user.account)"
                url += "&password=\(user.passwordMD5WithRand)"
                url += "&\(user.config!.sessionName)"
                url += "=\(user.config!.sessionID)"
                url += "&t=\(viewType)"
                return url;
            }
            
            url += "m=\(moduleName)&f=\(methodName)"
            
            if moduleNameIdentifier == "api" {
                if methodNameIdentifier == "mobilegetlist" {
                    url += "&type=" + (params["type"] ?? "full")
                    url += "&object=" + (params["object"] ?? "all")
                    url += "&range=" + (params["range"] ?? "0")
                    url += "&last=" + (params["last"] ?? "")
                    url += "&records=" + (params["records"] ?? "all")
                    url += "&format=" + (params["format"] ?? "index");
                } else if methodNameIdentifier == "mobilegetinfo" {
                    url += "&id=" + params["id"]! + "&type=" + params["type"]!
                }
            }
            
            url += "&type=\(viewType)&\(user.config!.sessionName!)=\(user.config!.sessionID!)"
        }
        
        return url;
    }
    
    static func getConfig(address: String, complete: ((ZentaoConfig?) -> Void)) {
        HTTP.get(address + "/index.php?mode=getconfig", complete: {(json: JSON?) -> Void in
            if let jsonObj = json {
                complete(ZentaoConfig(json:jsonObj))
            } else {
                complete(nil)
            }
        })
    }
    
    static func login(user: User, complete: ((result: Bool, error: String, message: String) -> Void)) {
        getConfig(user.address) {
            if let config = $0 {
                user.config = config
                if config.isZentaoAPIAvailable {
                    HTTP.get(self.concatUrl(["module": "user", "method": "login"], withUser: user), complete: {
                        (json: JSON?) -> Void in
                        if let jsonObj = json {
                            if jsonObj["status"].stringValue == "success" {
                                user.fromJSON(jsonObj["user"])
                                user.status = .Online
                                user.lastLoginTime = NSDate()
                                complete(result: true, error: "", message: "登录成功！")
                            } else {
                                complete(result: false, error: "WRONG_ACCOUNT", message: jsonObj["message"].string
                                    ?? "账号或密码错误。")
                            }
                        } else {
                            complete(result: false, error: "LOGIN_FAILED", message: "暂时无法登录到禅道服务器。")
                        }
                    })
                } else {
                    complete(result: false, error: "WRONG_VERSION", message: "此禅道版本不受支持，需要禅道专业版4.5及以上。")
                }
            } else {
                complete(result: false, error: "WRONG_CONNECT", message: "暂时无法连接到禅道服务器。")
            }
        }
    }
    
    static func getItem(entityType: EntityType, id: String, user: User, complete: ((result: Bool, jsonData: JSON?, message: String?) -> Void)) {
        let params = [
            "module": "api",
            "method": "mobileGetInfo",
            "type": entityType.name.lowercaseString,
            "id": id]
        HTTP.get(self.concatUrl(params, withUser: user)) {
            (json: JSON?) in
            if let jsonObj = json {
                if jsonObj["status"].stringValue == "success" {
                    let jsonData = jsonObj["data"]
                    if jsonData.isNullOrEmpty {
                        complete(result: true, jsonData: nil, message: "Success, but no new data there.")
                    } else {
                        complete(result: true, jsonData: jsonData, message: nil)
                    }
                } else {
                    complete(result: false, jsonData: nil, message: "The response from zentao server return in a wrong format.")
                }
                return
            }
            complete(result: false, jsonData: nil, message: "Can't connect to zentao server.")
        }
    }
    
    static func getItem(entityType: EntityType, id: Int, user: User, complete: ((result: Bool, jsonData: JSON?, message: String?) -> Void)) {
        getItem(entityType, id: String(id), user: user, complete)
    }
    
    static func getItemList(entityType: EntityType, user: User,
        options: (type: String, range: Int, records: Int, format: String), complete:
        ((result: Bool, jsonData: JSON?, message: String?) -> Void)) {
            
            let params = [
                "module": "api",
                "method": "mobileGetList",
                "object": entityType == .Default ? "all" : entityType.name.lowercaseString,
                "type": options.type,
                "range": String(options.range),
                "last": !user.isNeverSynced ? String(Int(round(user.lastSyncTime!.timeIntervalSince1970))) : "0",
                "records": String(options.records),
                "format": String(options.format)]
            
            HTTP.get(self.concatUrl(params, withUser: user)) {
                (json: JSON?) in
                if let jsonObj = json {
                    if jsonObj["status"].stringValue == "success" {
                        let jsonData = jsonObj["data"]
                        if jsonData.isNullOrEmpty {
                            complete(result: true, jsonData: nil, message: "Success, but no new data there.")
                        } else {
                            complete(result: true, jsonData: jsonData, message: nil)
                        }
                    } else {
                        complete(result: false, jsonData: nil, message: "The response from zentao server return in a wrong format.")
                    }
                    return
                }
                complete(result: false, jsonData: nil, message: "Can't connect to zentao server.")
            }
    }
    
    static func getItemList(entityType: EntityType, user: User, complete:
        ((result: Bool, jsonData: JSON?, message: String?) -> Void)) {
            getItemList(entityType, user: user, options: ("increment", 0, 1000, "index"), complete)
    }
    
    static func getItemList(user: User, complete:
        ((result: Bool, jsonData: JSON?, message: String?) -> Void)) {
            getItemList(.Default, user: user, options: ("increment", 0, 1000, "index"), complete)
    }
}
